Make JIT Serialization support arbitrary std::function<> IO

jjlilley opened the pull request:
Summary:
Right now, torch::save() uses std::ostream, which results in unnecessary
data copies in practice. Similar for torch::load().

Adding a std::function<size_t(const void*, size_t)> as an output option,
parallel to the existing filename and std::ostream apis, gives users the
flexibility to emit directly to a backing store.

For a simple case of appending the output to a std::string, we observe
significant benchmark savings (on order of -50%), even with the
minor std::function<> dispatch overhead. The main reason is that
std::ostringstream effectively requires 2 extra copies of the data
beyond a simple string.append lambda.

We also provide a parallel api for the load(), though this one is
slightly more complex due to the need to do arbitrary position reads.

resistor commented on caffe2/serialize/inline_container.cc:
```diff
@@ -232,44 +232,73 @@ PyTorchStreamReader::~PyTorchStreamReader() {
   valid("closing reader for archive ", archive_name_.c_str());
 }
 
-size_t ostream_write_func(void *pOpaque, mz_uint64 file_ofs, const void *pBuf, size_t n) {
```
I'm not going to block on this, but it's preferable to separate unrelated formatting changes into their own diffs.

resistor commented on caffe2/serialize/inline_container.h:
```diff
@@ -141,18 +149,25 @@ class CAFFE2_API PyTorchStreamWriter final {
   std::unique_ptr<mz_zip_archive> ar_;
   std::string archive_name_;
-  std::ostream* out_;
+  std::ostream* out_ = nullptr;
```
Is it really providing any value to have both paths?  Could we not just convert all callers to use the `std::function` path?

jjlilley replied:
That sounds good, I'll unify behind std::function<>.

resistor commented on test/cpp/api/serialize.cpp:
```diff
@@ -60,6 +60,25 @@ TEST(SerializeTest, BasicToFile) {
+  });
+  std::istringstream readHelper(serialized);
+  torch::Tensor y;
+  torch::load(y, readHelper);
```
For API orthogonality, we should probably make the same changes to `load` as well.

jjlilley replied:
Yes, totally agree.
My intention was partly to gauge whether there would be objections that I hadn't thought of to this type of approach before doing the parallel load path. But I'll go ahead and change load as well.

zdevito commented:
This is a good idea, Ill let Owen mark when it is ready to land.

jjlilley commented:
I just uploaded an updated version:
 1) There is std::function<> support for both torch::load() and torch::save() codepaths now
 2) I did simplify/unify the lowest layer PyTorchStreamWriter impl a bit, but chose not to remove the filename-based constructor api because of the python binding to that constructor.

resistor commented on caffe2/serialize/inline_container.h:
```diff
@@ -123,11 +123,15 @@ class CAFFE2_API PyTorchStreamReader final {
-  : PyTorchStreamWriter("archive", out) {}
-
-  void writeRecord(const std::string& name, const void* data, size_t size, bool compress = false);
+  explicit PyTorchStreamWriter(std::string archive_name);
```
Can't this just become a wrapper around the `std::function` constructor?

resistor commented on caffe2/serialize/inline_container.h:
```diff
@@ -141,13 +145,16 @@ class CAFFE2_API PyTorchStreamWriter final {
-  std::ostream* out_;
-  std::ofstream file_stream_;
+  std::string archive_name_plus_slash_;
+  std::unique_ptr<std::ofstream> file_stream_;
```
Why does this need to be a `unique_ptr`?  I beliefe `ofstream` can be default constructed.

jjlilley replied:
It didn't need to be a unique_ptr<>, was mostly trying to save work in the case where it's not needed. Reverted this part back.

jjlilley commented:
(also fixed a few issues to make clang-tidy happy)

jjlilley commented:
Looked into the failing tests - related to a bug in c10/util/tempfile.h, which was constructing a temp file name with a trailing '\0' in the std::string. Uploading the fix.

Commit:
Make JIT Serialization support arbitrary std::function<> I/O (#27586)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/27586

Right now, torch::save() uses std::ostream, which results in unnecessary
data copies in practice. Similar for torch::load().

Adding a std::function<size_t(const void*, size_t)> as an output option,
parallel to the existing filename and std::ostream apis, gives users the
flexibility to emit directly to a backing store.

For a simple case of appending the output to a std::string, we observe
significant benchmark savings (on order of -50%), even with the
minor std::function<> dispatch overhead. The main reason is that
std::ostringstream effectively requires 2 extra copies of the data
beyond a simple string.append lambda.

We also provide a parallel api for the load(), though this one is
slightly more complex due to the need to do arbitrary position reads.

Test Plan:
buck test mode/dev-nosan caffe2/test/...
      (Basic serialization test in caffe2/test/cpp/api/serialize.cpp)
      Benchmark in experimental/jeremyl/c2/SerializationBench.cpp, with D17823443
        (1M time goes from 90ms -> 40ms, albeit with crc patch applied)

Differential Revision: D17822962

fbshipit-source-id: 2cf2b24ad7d4212f2e4a49d7abc257cbdf33d206

jjlilley commented:
Thanks for looking at this!

yf225 commented:
It seems that this PR is breaking macOS builds on master:
```
Oct 10 15:32:13 ../torch/csrc/api/src/serialize/input-archive.cpp:115:22: error: no matching function for call to 'min'
Oct 10 15:32:13       size_t nread = std::min(pos + n, size_) - pos;
Oct 10 15:32:13                      ^~~~~~~~
```

yf225 commented:
Sorry that to unbreak master, I will have to revert this PR.

yf225 commented:
We'll need to look at the Github CI on this PR - specifically `pytorch_macos_10_13_py3_build` and `pytorch_macos_10_13_cuda9_2_cudnn7_py3_build` was failing.
